src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -532,37 +528,34 @@ void G1CollectedHeap::iterate_regions_in_range(MemRegion range, const Func& func
   }
 }
 
-bool G1CollectedHeap::alloc_archive_regions(MemRegion range) {
+HeapWord* G1CollectedHeap::alloc_archive_region(size_t word_size, HeapWord* preferred_addr) {
   assert(!is_init_completed(), "Expect to be called at JVM init time");
   MutexLocker x(Heap_lock);
 
   MemRegion reserved = _hrm.reserved();
 
+  if (reserved.word_size() <= word_size) {
+    log_info(gc, heap)("Unable to allocate regions as archive heap is too large; size requested = " SIZE_FORMAT
+                       " bytes, heap = " SIZE_FORMAT " bytes", word_size, reserved.word_size());
+    return nullptr;
+  }
+
   // Temporarily disable pretouching of heap pages. This interface is used
   // when mmap'ing archived heap data in, so pre-touching is wasted.
   FlagSetting fs(AlwaysPreTouch, false);
 
-  // For the specified MemRegion range, allocate the corresponding G1
-  // region(s) and mark them as old region(s).
-  HeapWord* start_address = range.start();
-  size_t word_size = range.word_size();
-  HeapWord* last_address = range.last();
   size_t commits = 0;
-
-  guarantee(reserved.contains(start_address) && reserved.contains(last_address),
-            "MemRegion outside of heap [" PTR_FORMAT ", " PTR_FORMAT "]",
-            p2i(start_address), p2i(last_address));
-
-  // Perform the actual region allocation, exiting if it fails.
-  // Then note how much new space we have allocated.
+  // Attempt to allocate towards the end of the heap.
+  HeapWord* start_addr = reserved.end() - align_up(word_size, HeapRegion::GrainWords);
+  MemRegion range = MemRegion(start_addr, word_size);
+  HeapWord* last_address = range.last();
```

Comment:
Do I understand this - and the old code - correctly: we map the archive region at the end of the heap, therefore, if heap size changed between dump time and runtime, we will always have to relocate? I didn't look but I assume the dumptime heap size is the default heap size?

Another, possibly stupid, question: why don't we use the Heap bottom - we'd still avoid fragmentation but have a much higher chance of not relocating?

Reply:
> we map the archive region at the end of the heap, therefore, if heap size changed between dump time and runtime, we will always have to relocate

Not always. For small heap sizes under 2G with compressedoops the heap is mapped such that top of the heap is at 4G boundary. So by keeping the archive regions towards the end, we can get same offset even if the heap size changes. If we map towards the bottom, then the offsets would change as the heap size changes.

For example, the default archive is created with 128m. So executing:
```
$ java -Xlog:cds -Xmx128m -version
[0.003s][info][cds] The current max heap size = 128M, HeapRegion::GrainBytes = 1048576
[0.003s][info][cds]     narrow_klass_base = 0x0000000800000000, narrow_klass_shift = 0
[0.003s][info][cds]     narrow_oop_mode = 0, narrow_oop_base = 0x0000000000000000, narrow_oop_shift = 0
[0.003s][info][cds]     heap range = [0x00000000f8000000 - 0x0000000100000000]
[0.003s][info][cds] Preferred address to map heap data (to avoid relocation) is 0x00000000ffe00000
[0.003s][info][cds] Heap data mapped at 0x00000000ffe00000, size =  1071752 bytes
[0.003s][info][cds] CDS heap data relocation delta = 0 bytes
```

Changing the heap size to 256m:
```
$ java -Xlog:cds -Xmx256m -version
[0.003s][info][cds] The current max heap size = 256M, HeapRegion::GrainBytes = 1048576
[0.003s][info][cds]     narrow_klass_base = 0x0000000800000000, narrow_klass_shift = 0
[0.003s][info][cds]     narrow_oop_mode = 0, narrow_oop_base = 0x0000000000000000, narrow_oop_shift = 0
[0.003s][info][cds]     heap range = [0x00000000f0000000 - 0x0000000100000000]
[0.003s][info][cds] Preferred address to map heap data (to avoid relocation) is 0x00000000ffe00000
[0.003s][info][cds] Heap data mapped at 0x00000000ffe00000, size =  1071752 bytes
[0.003s][info][cds] CDS heap data relocation delta = 0 bytes
```

Notice that relocation delta is 0 in both cases and the heap data is mapped at same address.

Reply:
> I didn't look but I assume the dumptime heap size is the default heap size?

I am wondering why do you think dumptime heap size is the default heap size?  When creating an archive the heap size can be anything. For default archive it is 128m.

Reply:
> > I didn't look but I assume the dumptime heap size is the default heap size?
> 
> 
> 
> I am wondering why do you think dumptime heap size is the default heap size?  When creating an archive the heap size can be anything. For default archive it is 128m.

The dump time heap size is set to 128mb for the default archive so that it can be mapped without relocation for small workloads with heaps that can fit under the 4gb boundary.